consensus/src/dag/dag_store.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, ensure};
+use aptos_consensus_types::common::{Author, Payload, Round};
+use aptos_crypto::{
+    hash::{CryptoHash, CryptoHasher},
+    HashValue,
+};
+use aptos_crypto_derive::CryptoHasher;
+use aptos_types::{aggregate_signature::AggregateSignature, validator_verifier::ValidatorVerifier};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    ops::Deref,
+    sync::Arc,
+};
+
+/// Represents the metadata about the node, without payload and parents from Node
+#[derive(Clone, Serialize, Deserialize)]
+pub struct NodeMetadata {
+    epoch: u64,
+    round: Round,
+    author: Author,
+    timestamp: u64,
+    digest: HashValue,
+}
+
+/// Node representation in the DAG, parents contain 2f+1 strong links plus weak links
+#[derive(Clone, Serialize, Deserialize, CryptoHasher)]
+pub struct Node {
+    metadata: NodeMetadata,
+    payload: Payload,
+    parents: Vec<NodeMetadata>,
+}
+
+impl Node {
+    pub fn new(
+        epoch: u64,
+        round: Round,
+        author: Author,
+        timestamp: u64,
+        payload: Payload,
+        parents: Vec<NodeMetadata>,
+    ) -> Self {
+        let digest = Self::calculate_digest(epoch, round, author, timestamp, &payload, &parents);
+
+        Self {
+            metadata: NodeMetadata {
+                epoch,
+                round,
+                author,
+                timestamp,
+                digest,
+            },
+            payload,
+            parents,
+        }
+    }
+
+    /// Calculate the node digest based on all fields in the node
+    fn calculate_digest(
+        epoch: u64,
+        round: Round,
+        author: Author,
+        timestamp: u64,
+        payload: &Payload,
+        parents: &Vec<NodeMetadata>,
+    ) -> HashValue {
+        #[derive(Serialize)]
+        struct NodeWithoutDigest<'a> {
+            epoch: u64,
+            round: Round,
+            author: Author,
+            timestamp: u64,
```

Comment:
Do you need time stamp to make the digest unique?

Reply:
not really, it's used for producing the block timestamp

Reply:
May be get rid of this in the digest calculation in that case?

Reply:
why? this is part of the signed information, otherwise malicious node could reply the same digest with a different timestamp